src/hooks/use-dropdown-menu.tsx

```diff
@@ -0,0 +1,150 @@
+   const itemRefs = useRef([...Array(itemCount)].map(() => createRef<HTMLAnchorElement>()));
+   
+   
+   // Create type guard
+   const isKeyboardEvent = (e: React.KeyboardEvent | React.MouseEvent): e is React.KeyboardEvent => (e as React.KeyboardEvent).key !== undefined;
+   
+   
+   // Handle changes in the state of the menu
+   useEffect(() => {
+       // Stop if the ref currently points to a null value
+       if (!itemRefs.current[0].current || !buttonRef.current) {
+           return;
+       }
+
+
+       // Stop if this is the first fire of the Hook, and update the ref
+       if (firstRun.current) {
+           firstRun.current = false;
+           return;
+       }
+
+       
+       // If the menu is currently open focus on the first item in the menu, else return focus to the button if state indicates that is desired behavior
+       if (isOpen) {
+           itemRefs.current[0].current.focus();
+           moveFocus(0);
+       } else if (returnFocusToButton) {
+           buttonRef.current.focus();
+       }
+   }, [isOpen, returnFocusToButton]);
+
+
+   // Handles moving the focus between menu items
+   const moveFocus = (itemIndex: number) => {
+        currentFocusIndex.current = itemIndex
+        itemRefs.current[itemIndex].current?.focus();
+   }
+   
+
+   // Create a handler function for the button's clicks and keyboard events
+   const buttonListener = (e: React.KeyboardEvent | React.MouseEvent) => {
+       // Detect if event was a keyboard event or a mouse event
+       if (isKeyboardEvent(e)) {
+           const { key } = e;
+           
+           if (!(key === 'Tab' || key === 'Shift')) {
+               e.preventDefault();
+           }
+           if (key === 'Enter' || key === ' ') {
+               setIsOpen(true);
+               setReturnFocusToButton(false);
+           } else if (key === 'Tab') {
+               setIsOpen(false);
+               setReturnFocusToButton(false);
+           }
+       } else {
+           setIsOpen(!isOpen);
+       }
+   };
+   
+   
+   // Create a function that handles menu logic based on keyboard events that occur on menu items
+   const itemListener = (e: React.KeyboardEvent<HTMLAnchorElement>) => {
+       // Create mutable value that initializes as the currentFocusIndex value
+       let newFocusIndex = currentFocusIndex.current;
+       
+
+       // Destructure the key property from the event object
+       const { key } = e;
+       
+       
+       // Prevent default browser behavior except in cases where maintaining the natural tab order is desired
+       if (!(key === 'Tab' || key === 'Shift' || key === 'Enter')) {
+           e.preventDefault();
+       }
+       
+       
+       // Controls whether the menu is open or closed, if the button should regain focus on close, and if a handler function should be called
+       if (key === 'Escape') {
+           setIsOpen(false);
+           setReturnFocusToButton(true);
```

Comment:
This is the only place where you're calling `setReturnFocusToButton(true);` now. Everywhere else you set it to `false`.

Setting the state causes the entire component to rerender, and upon rerendering, it moves the focus back to the button. That's an indirect way to simply move the focus. Why not just remove this state entirely and replace this line with:

```js
buttonRef.current.focus();
```